litr/src/main/java/com/linkedin/android/litr/utils/CodecUtils.java

```diff
@@ -172,99 +175,129 @@ public static int getHighestSupportedProfile(@NonNull String mimeType, boolean i
-        if (codecName != null) {
-            mediaCodec = MediaCodec.createByCodecName(codecName);
-        }
-        if (mediaCodec != null) {
-            mediaCodec.configure(mediaFormat, surface, null, isEncoder ? MediaCodec.CONFIGURE_FLAG_ENCODE : 0);
+    @NonNull
+    private static MediaCodec getAndConfigureCodec(
+            @NonNull MediaFormat mediaFormat,
+            @Nullable Surface surface,
+            boolean isEncoder,
+            @NonNull List<Callable<MediaCodec>> supportedMediaCodecs) throws IllegalStateException, IOException {
+
+        MediaCodec mediaCodec = null;
+        IOException error = null;
+        for (Callable<MediaCodec> callable : supportedMediaCodecs) {
+            try {
+                mediaCodec = callable.call();
+                if (mediaCodec != null) {
+                    mediaCodec.configure(mediaFormat, surface, null, isEncoder ? MediaCodec.CONFIGURE_FLAG_ENCODE : 0);
+                    break;
+                }
+            } catch (Exception e) {
+                if (mediaCodec != null) {
+                    mediaCodec.release();
+                    mediaCodec = null;
+                }
+                if (e instanceof IOException) {
+                    error = (IOException) e;
+                }
+            }
         }
 
+        if (mediaCodec == null) {
+            if (error != null) {
+                throw error;
+            } else {
+                throw new IllegalStateException();
+            }
+        }
         return mediaCodec;
     }
 
+    /**
+     * This will iterate over all available codecs that support the mimeType param, and return them as a list of
+     * {@link Callable Callable#MediaCodec}, when will create the codec once calling {@link Callable#call()}
+     * @param encoder flag indicating if encoder codec is requested
+     * @param mimeType the mime type for which to get the codecs that support it
+     * @param mediaFormat {@link MediaFormat} if non-null, then only get the codecs that support the mediaFormat
+     */
+    @NonNull
+    private static List<Callable<MediaCodec>> findCodecForFormatOrType(boolean encoder, @NonNull String mimeType,
+            @Nullable MediaFormat mediaFormat) {
+
+        List<Callable<MediaCodec>> supportedMediaCodecs = new ArrayList<>();
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            MediaCodecList mediaCodecList = new MediaCodecList(MediaCodecList.ALL_CODECS);
+            for (MediaCodecInfo info : mediaCodecList.getCodecInfos()) {
+                if (info.isEncoder() != encoder) {
+                    continue;
+                }
+                try {
+                    MediaCodecInfo.CodecCapabilities caps = info.getCapabilitiesForType(mimeType);
+                    if (caps != null && (mediaFormat == null || caps.isFormatSupported(mediaFormat))) {
+                        supportedMediaCodecs.add(() -> MediaCodec.createByCodecName(info.getName()));
```

Comment:
Just curious, why are you using `Callable`?

Reply:
I think it could be better to not create multiple instances of `Mediacodec`, where most of the time I'll use the first one, and I'll need to releases all other created instances that unneeded. so I used the `Callable` to create `MediaCodec` instance when needed only.

Reply:
Very clever. :)